### PR TITLE
Add Multiple Argument Functionality to STLC Interface

### PR DESCRIPTION
**Pull request by agureev**
This commit adds following functionality:

1) `lamb` now accepts a list of types for its `tdom` accessor allowing for functions to be defined in a complex context. The domain of such a lambda term will then be the iterated product.

2) 'app' now accepts a list of terms for its `term` accessor allowing for a multi-argument function to be iteratively applied.

3) Change `index` functionality to follow usual convention of reading contexts from left to right.

4) Change the STLC -> GEB compilation to be uncurried solving #106 and allowing for a neater solution to #90

### CODE REVIEW

**Comment by agureev** [comment]
Adding @rokopt as a new reviewer to receive approval on the new compilation pass. Note that the application interpretation follows from composition-as-substitution definition in the term model.

File: src/util/utils.lisp [hunk 1]
```diff
@@ -201,3 +201,13 @@ Further this can be used in type signatures
...
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;               Additional Utils
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+
+(defun apply-n (times f initial)
+  (let ((value initial))
+    (dotimes (n times value)
+      (setf value (funcall f value)))))
```
Comment (mariari) [inline src/util/utils.lisp:213]: Probably would good to have a doc string with this

File: src/lambda/lambda.lisp [hunk 2]
```diff
@@ -55,8 +55,8 @@ length at least [i] and then produces the [i]'th entry of the context where
 contexts are read right to left."
   (let ((l (length ctx)))
     (if (< i l)
-        (nth (- l (+ i 1)) ctx)
-        (error "argument exceeds length of context"))))
+        (nth i ctx)
+        (error "Argument exceeds length of context"))))
```
Comment (mariari) [inline src/lambda/lambda.lisp:59]: Yeah much nicer

File: src/lambda/lambda.lisp [hunk 3]
```diff
@@ -121,10 +121,10 @@ functions taking in context and terms below as well."))
                                (if (typep type-of-term 'prod)
                                    (snd ann-term :ttype (mcadr type-of-term))
                                    (error "type of term not of product type"))))
-        ((lamb tdom term)   (let ((ant (ann-term1 (cons tdom ctx) term)))
+        ((lamb tdom term)   (let ((ant (ann-term1 (append tdom ctx) term)))
```
Comment (mariari) [inline src/lambda/lambda.lisp:124]: This causes tests to fail, as tdom is still so1 in some situations

Comment (mariari) [reply src/lambda/lambda.lisp:124]: With the merged commits this is no longer an issue

File: src/lambda/lambda.lisp [hunk 4]
```diff
@@ -121,27 +121,26 @@ functions taking in context and terms below as well."))
...
-                                 (ann-term1 ctx term)
+                                 (mapcar (lambda (trm) (ann-term1 ctx trm)) term)
                                  :ttype (hom-cod ctx fun)))
         ((index pos)        (index pos
                                    :ttype (index-check pos ctx)))
         ((case-on on ltm rtm)
          (let* ((ann-on     (ann-term1 ctx on))
                 (type-of-on (ttype ann-on))
-                (ann-left   (ann-term1 (cons (mcar type-of-on) ctx) ltm))
+                (ann-left   (ann-term1 (cons  (mcar type-of-on) ctx) ltm))
```
Comment (mariari) [inline src/lambda/lambda.lisp:136]: Seems like a needless change that makes formatting worse

File: src/lambda/trans.lisp [hunk 5]
```diff
@@ -85,17 +85,14 @@ produces an error. Error of such kind mind pop up both on the level of evaluatin
...
                ((app fun term)
-                (let ((tofun (ttype fun)))
-                  (comp
-                   (so-eval (fun-to-hom (mcar tofun))
-                            (fun-to-hom (mcadr tofun)))
-                   (geb:pair (rec context
-                                  fun)
-                             (rec context
-                                  term)))))
+                (comp  (rec context fun)
```
Comment (mariari) [inline src/lambda/trans.lisp:90]: change

```lisp
(comp  (rec context fun) ...)
```

into

```lisp
(comp (rec context fun) ...)

File: src/specs/lambda.lisp [hunk 6]
```diff
@@ -333,16 +333,37 @@ where we can possibly include type information by
...
-THe intended semnatics are: [TDOM][generic-function] is the type (and
-hence a [SUBSTOBJ][GEB.SPEC:SUBSTOBJ]) which is the domain of the
-function type. [TERM][generic-function] is a term (and hence an
-[STLC][type]) of the codomain of the function type gotten in the
-context of the domain.
+The intended semnatics are: [TDOM][generic-function] is a list of types (and
+hence a list of [SUBSTOBJ][GEB.SPEC:SUBSTOBJ]) whose iterative product of
+components form the domain of the function type. [TERM][generic-function]
+is a term (and hence an [STLC][type]) of the codomain of the function type
+gotten in the context to whom we append the list of the domains.
```
Comment (mariari) [inline src/specs/lambda.lisp:340]: Some of these docs need to propagate into the class docs

**Review by mariari — CHANGES_REQUESTED** [review]
I'd like to see updated docs on the classes. Since the semantics of lambda and app changed, I wish for the class documentation to be updated to reflect this, and type signatures added to the classes and constructors as well, so I can see them at a glance.

Overall after some docs changes I did manually the pr seems to work, I want the HOF example put into a test as well

**Review by rokopt — APPROVED** [review]
This makes sense to me -- thank you!

As a general comment:  I suggest we ask the Juvix team how much it'll take for them to adapt to the change to the indexing; if it's at all difficult then we could provide a utility that just syntactically transforms a program written with the old indexing to one written with the new.

**Pull request merged**